### PR TITLE
Use jetbrains annotations and remap mc to use them.

### PR DESCRIPTION
**Pull request by modmuss50**
This would need to go into a 0.5 loom as it does have a slim chance of breaking as some people included jsr in their buildscript to work around some compiler oddities.

### CODE REVIEW

File: src/main/java/net/fabricmc/loom/providers/MinecraftMappedProvider.java [hunk 1]
```diff
@@ -118,8 +122,15 @@ private void mapMinecraftJar() throws IOException {
 	}
 
 	public TinyRemapper getTinyRemapper(String fromM, String toM) throws IOException {
+		TinyTree annotationMappingsTree;
+
+		try (BufferedReader reader = new BufferedReader(new InputStreamReader(MinecraftMappedProvider.class.getClassLoader().getResourceAsStream("jsrToJetbrains.tiny")))) {
```
Comment (natanfudge) [inline src/main/java/net/fabricmc/loom/providers/MinecraftMappedProvider.java]: I would have done 
```
Map<String,String> jsrToJetbrains = new HashMap {{
   put("javax/annotation/Nullable", "org/jetbrains/annotations/Nullable");
   ...
}};
```
And then
```
.withMappings(acceptor -> jsrToJetbrains.forEach((jetbrains,jsr) -> acceptor.acceptClass(jetbrains,jsr)))
```

**Review by natanfudge — CHANGES_REQUESTED** [review]

File: src/main/java/net/fabricmc/loom/providers/MinecraftMappedProvider.java (hunk 1 above)
Comment (modmuss50) [reply src/main/java/net/fabricmc/loom/providers/MinecraftMappedProvider.java]: Yes, thats prob better. I expected there to be more to remap possibly including methods.

**Review by Player3324 — APPROVED** [review]

**Review by shedaniel — APPROVED** [review]

**Comment by liach** [comment]
Imo this should really be done on yarn's side and annotation location needs adjustment as well. Suffices as a bridging practice though.

**Review by liach — APPROVED** [review]

**Review by Earthcomputer — APPROVED** [review]

File: src/main/java/net/fabricmc/loom/providers/MinecraftMappedProvider.java [hunk 2]
```diff
@@ -40,6 +42,12 @@
 import net.fabricmc.loom.util.DependencyProvider;
 
 public class MinecraftMappedProvider extends DependencyProvider {
+	private static final Map<String, String> jsrToJetbrains = new ImmutableMap.Builder<String, String>()
```
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/providers/MinecraftMappedProvider.java]:
```suggestion
	private static final Map<String, String> JSR_TO_JETBRAINS = new ImmutableMap.Builder<String, String>()
```

**Review by Juuxel — CHANGES_REQUESTED** [review]

File: src/main/java/net/fabricmc/loom/providers/MinecraftMappedProvider.java (hunk 1 above)
Comment (JamiesWhiteShirt) [reply src/main/java/net/fabricmc/loom/providers/MinecraftMappedProvider.java]: Every time you use double brace initialization, god kills a kitten

File: src/main/java/net/fabricmc/loom/providers/MinecraftMappedProvider.java (hunk 2 above)
Comment (liach) [reply src/main/java/net/fabricmc/loom/providers/MinecraftMappedProvider.java]: guess so, this map is effectively immutable

**Review by natanfudge — APPROVED** [review]

**Review by ramidzkh — APPROVED** [review]

**Comment by liach** [comment]
@ramidzkh why is it bad to do it on yarn's side? these nullability preferably should be moved to type annotations' locations, and we should really make yarn a documentation system that document extra annotations that *should have been present* in the code.

**Comment by Earthcomputer** [comment]
> that document extra annotations that should have been present in the code

and also change the annoying ones like `@Nullable ClientPlayerEntity player;` and `@Deprecated` from those Block methods (replace with `@OverrideOnly`)

**Comment by i509VCB** [comment]
Really the `@OverrideOnly` handling should be part of yarn imo. Someone should write an RFC for adding extra annotations to tiny spec for use in yarn or however else they could see it working. I am 100% sure you will have to manually catalog all methods and fields that.

> `@Nullable ClientPlayerEntity player;`

That field in MinecraftClient has the intended annotation. The ClientPlayerEntity is only present if in game. Some javadoc for yarn specifying when those nullable fields are available would be nice to add.

**Review by Pyrofab — APPROVED** [review]

**Comment by Earthcomputer** [comment]
> That field in MinecraftClient has the intended annotation

But it's annoying to have the IDE warning because most of the time you're using it you know that the player exists.

**Comment by i509VCB** [comment]
I asusme you know you can suppress the warning.

**Pull request merged**